### PR TITLE
feat: Add functions to retrieve event payload(s)

### PR DESCRIPTION
**Pull request by nwithan8**
# Description

- Add function to retrieve payload/payloads for a given Event
- Add unit tests, cassettes
- Add new Payload class, fix Event class schema

# Testing

- New unit tests, cassettes

# Pull Request Type

Please select the option(s) that are relevant to this PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Improvement (fixing a typo, updating readme, renaming a variable name, etc)

### CODE REVIEW

File: EasyPost.Tests/ServicesTests/EventServiceTest.cs [hunk 1]
```diff
@@ -54,6 +67,65 @@ public async Task TestRetrieve()
...
+            // Create a webhook to receive the event
+            const string url = "https://example.com";
+            Webhook webhook = await Client.Webhook.Create(new Dictionary<string, object> { { "url", url } });
+            CleanUpAfterTest(webhook.Id);
+
+            // Create a batch to trigger an event
+            Batch _ = await Client.Batch.Create(new Dictionary<string, object> { { "shipments", new List<Dictionary<string, object>> { Fixtures.BasicShipment } } });
+
+            // Wait for the event to be created
+            await Task.Delay(5000);
```
Comment (Justintime50) [inline EasyPost.Tests/ServicesTests/EventServiceTest.cs:86]: Please wrap this in the same logic we use elsewhere to ensure it only happens on record.

File: EasyPost.Tests/ServicesTests/EventServiceTest.cs [hunk 2]
```diff
@@ -54,6 +67,65 @@ public async Task TestRetrieve()
...
+
+            // Retrieve all events and extract the newest one
+            EventCollection eventCollection = await Client.Event.All(new Dictionary<string, object> { { "page_size", Fixtures.PageSize } });
+            Event @event = eventCollection.Events[0];
+
+            // Retrieve the payloads for the event
+            List<Payload> retrievedPayloads = await Client.Event.RetrievePayloadsForEvent(@event);
```
Comment (Justintime50) [inline EasyPost.Tests/ServicesTests/EventServiceTest.cs:93]: Nit: Could just be `payloads` heh

File: EasyPost.Tests/ServicesTests/EventServiceTest.cs [hunk 3]
```diff
@@ -54,6 +67,65 @@ public async Task TestRetrieve()
...
+            // Create a webhook to receive the event
+            const string url = "https://example.com";
+            Webhook webhook = await Client.Webhook.Create(new Dictionary<string, object> { { "url", url } });
+            CleanUpAfterTest(webhook.Id);
+
+            // Create a batch to trigger an event
+            Batch _ = await Client.Batch.Create(new Dictionary<string, object> { { "shipments", new List<Dictionary<string, object>> { Fixtures.BasicShipment } } });
+
+            // Wait for the event to be created
+            await Task.Delay(5000);
```
Comment (Justintime50) [inline EasyPost.Tests/ServicesTests/EventServiceTest.cs:114]: Same as above for record only

File: EasyPost.Tests/ServicesTests/EventServiceTest.cs [hunk 4]
```diff
@@ -54,6 +67,65 @@ public async Task TestRetrieve()
...
+            Event @event = eventCollection.Events[0];
+
+            // Retrieve a specific payload for the event
+            // Payload does not exist due to queueing, so this will throw an exception
+            // invalid payload should throw an exception
+            await Assert.ThrowsAsync<NotFoundError>(async () => await Client.Event.RetrievePayloadForEvent(@event, "payload_11111111111111111111111111111111"));
+
+            // Invalid payload ID length will throw a 500, library remaps this to a 422
+            await Assert.ThrowsAsync<InvalidRequestError>(async () => await Client.Event.RetrievePayloadForEvent(@event, "payload_123"));
+        }
```
Comment (Justintime50) [inline EasyPost.Tests/ServicesTests/EventServiceTest.cs:127]: Rip this logic and test out.

File: EasyPost/Services/EventService.cs [hunk 5]
```diff
@@ -42,6 +43,35 @@ internal EventService(EasyPostClient client)
...
+        {
+            try
+            {
+                return await Get<Payload>($"events/{@event.Id}/payloads/{payloadId}");
+            }
+            catch (InternalServerError)
+            {
+                // The API returns a 500 error if the payload ID is malformed.
+                // This is a catch clause to instead throw a more appropriate exception.
+                throw new InvalidRequestError($"Payload ID is invalid length.", 422);
```
Comment (Justintime50) [inline EasyPost/Services/EventService.cs:71]: Please remove this logic as this wraps up all 500 errors to this incorrect error.

**Review by Justintime50 — DISMISSED** [review]

File: EasyPost.Tests/ServicesTests/EventServiceTest.cs [hunk 6]
```diff
@@ -54,6 +67,69 @@ public async Task TestRetrieve()
...
+            Event @event = eventCollection.Events[0];
+
+            // Retrieve a specific payload for the event
+            // Payload does not exist due to queueing, so this will throw an exception
+            // invalid payload should throw an exception
+            await Assert.ThrowsAsync<NotFoundError>(async () => await Client.Event.RetrievePayloadForEvent(@event, "payload_11111111111111111111111111111111"));
+
+            // Invalid payload ID length will throw a 500
+            await Assert.ThrowsAsync<InternalServerError>(async () => await Client.Event.RetrievePayloadForEvent(@event, "payload_123"));
+        }
```
Comment (Justintime50) [inline EasyPost.Tests/ServicesTests/EventServiceTest.cs:131]: Same concern here as in other PRs. We shouldn't be testing this.

**Review by Justintime50 — CHANGES_REQUESTED** [review]
The comments here are too verbose as they are in other PRs.

File: EasyPost/Models/API/Payload.cs [hunk 7]
```diff
@@ -0,0 +1,13 @@
+namespace EasyPost.Models.API
+{
+    public class Payload : Event // A payload has the same schema as an Event
```
Comment (jchen293) [inline EasyPost/Models/API/Payload.cs:3]: I am not sure if this is true, according to the Go: https://github.com/EasyPost/easypost-go/blob/master/event.go#L35-L51

Comment (Justintime50) [reply EasyPost/Models/API/Payload.cs:3]: Oh yeah, this is definitely not true.

File: EasyPost/Services/EventService.cs [hunk 8]
```diff
@@ -42,6 +43,23 @@ internal EventService(EasyPostClient client)
         [CrudOperations.Read]
         public async Task<Event> Retrieve(string id) => await Get<Event>($"events/{id}");
 
+        /// <summary>
+        ///     Retrieve all <see cref="Payload"/>s for an <see cref="Event"/>.
+        /// </summary>
+        /// <param name="event"><see cref="Event"/> to retrieve payloads for.</param>
+        /// <returns>A list of <see cref="Payload"/> objects.</returns>
+        public async Task<List<Payload>> RetrieveAllPayloadsForEvent(Event @event) => await Get<List<Payload>>($"events/{@event.Id}/payloads", rootElement: "payloads");
```
Comment (jchen293) [inline EasyPost/Services/EventService.cs:51]: This function should be called `RetrieveAllPayloads` to match other CLs

File: EasyPost/Services/EventService.cs [hunk 9]
```diff
@@ -42,6 +43,23 @@ internal EventService(EasyPostClient client)
...
+
+        /// <summary>
+        ///     Retrieve a specific <see cref="Payload"/> for an <see cref="Event"/>.
+        /// </summary>
+        /// <param name="event"><see cref="Event"/> to retrieve payload for.</param>
+        /// <param name="payloadId">ID of payload to retrieve.</param>
+        /// <returns>A <see cref="Payload"/> object.</returns>
+        /// <exception cref="InvalidRequestError">Thrown if the specified payload ID is malformed.</exception>
+        /// <exception cref="NotFoundError">Thrown if the specified payload is not found.</exception>
+        public async Task<Payload> RetrievePayloadForEvent(Event @event, string payloadId) => await Get<Payload>($"events/{@event.Id}/payloads/{payloadId}");
```
Comment (jchen293) [inline EasyPost/Services/EventService.cs:61]: This function should be called `RetrievePayload` to match other CLs

File: EasyPost/Services/EventService.cs (hunk 8 above)
Comment (nwithan8) [reply EasyPost/Services/EventService.cs:51]: Gross, I thought we decided for the hybrid design we would be passing whole objects around in function, not just the object IDs.

But I can't find anything to support that (C# was rewritten not with the hybrid design, and Java and PHP that were rewritten with the hybrid design are passing in only the ID of the object its executing against).

I hate that, but I'll adhere to it to make it consistent.

**Review by jchen293 — APPROVED** [review]

**Pull request merged**